tracing/otel.go

```diff
@@ -0,0 +1,253 @@
+		cfg.samplerType = "remote"
+	}
+
+	// Parse tags
+	cfg.jaegerTags, err = parseJaegerTags(os.Getenv(envJaegerTags))
+	if err != nil {
+		return cfg, errors.Wrapf(err, "could not parse %s", envJaegerTags)
+	}
+	return cfg, nil
+}
+
+// initJaegerTracerProvider initializes a new Jaeger Tracer Provider.
+func (cfg config) initJaegerTracerProvider(serviceName string) (io.Closer, error) {
+	// Read environment variables to configure Jaeger
+	var ep jaegerotel.EndpointOption
+	// Create the jaeger exporter: address can be either agent address (host:port) or collector Endpoint.
+	if cfg.agentHostPort != "" {
+		ep = jaegerotel.WithAgentEndpoint(
+			jaegerotel.WithAgentHost(cfg.agentHost),
+			jaegerotel.WithAgentPort(cfg.agentPort))
+	} else {
+		ep = jaegerotel.WithCollectorEndpoint(
+			jaegerotel.WithEndpoint(cfg.collectorEndpoint))
+	}
+	exp, err := jaegerotel.New(ep)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure sampling strategy
+	sampler := tracesdk.AlwaysSample()
+	if cfg.samplerType == "const" {
+		if cfg.samplerParam == 0 {
+			sampler = tracesdk.NeverSample()
+		}
+	} else if cfg.samplerType == "probabilistic" {
+		tracesdk.TraceIDRatioBased(cfg.samplerParam)
+	} else if cfg.samplerType == "remote" {
+		sampler = jaegerremote.New(serviceName, jaegerremote.WithSamplingServerURL(cfg.samplingServerURL),
+			jaegerremote.WithInitialSampler(tracesdk.TraceIDRatioBased(cfg.samplerParam)))
+	} else if cfg.samplerType != "" {
+		return nil, errors.Errorf("unknown sampler type %q", cfg.samplerType)
+	}
+	customAttrs := cfg.jaegerTags
+	customAttrs = append(customAttrs,
+		attribute.String("samplerType", cfg.samplerType),
+		attribute.Float64("samplerParam", cfg.samplerParam),
+		attribute.String("samplingServerURL", cfg.samplingServerURL),
+	)
+	res, err := NewResource(serviceName, customAttrs)
+	if err != nil {
+		return nil, err
+	}
+
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithBatcher(exp),
+		tracesdk.WithResource(res),
+		tracesdk.WithSampler(sampler),
+	)
+
+	// Set TracerProvider with pyroscope profiling.
+	otel.SetTracerProvider(otelpyroscope.NewTracerProvider(tp))
+
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator([]propagation.TextMapPropagator{
+		// w3c Propagator is the default propagator for opentelemetry
+		propagation.TraceContext{}, propagation.Baggage{},
+		// jaeger Propagator is for opentracing backwards compatibility
+		jaegerpropagator.Jaeger{},
+	}...))
+	return &Closer{tp}, nil
+}
+
+type Closer struct {
+	*tracesdk.TracerProvider
+}
+
+func (c Closer) Close() error {
+	return c.Shutdown(context.Background())
+}
+
+// NewResource creates a new OpenTelemetry resource using the provided service name and custom attributes.
+// This resource will be used for creating both tracers and meters, enriching telemetry data with context.
+func NewResource(serviceName string, customAttributes []attribute.KeyValue) (*resource.Resource, error) {
```

Comment:
It's probably fine for first iteration that service.namespace and service.version  aren't here. https://opentelemetry.io/docs/specs/semconv/resource/#service  I haven't seen semconf.ServiceNamespace used in grafana code. (Something to be aware of anyway)

Reply:
[Tempo also doesn't have it](https://github.com/grafana/tempo/blob/a4289d96f02f88b42ee168265d17dab3b1efe7e5/cmd/tempo/main.go#L233-L239), and they're the tracing experts for me.